codalab/worker/file_util.py

```diff
@@ -171,11 +184,192 @@ def do_unzip(filename):
+                tarFileName=parse_linked_bundle_url(path).bundle_uuid,
+                writeIndex=False,
+                clearIndexCache=False,
+                indexFileName=index_file.name,
+            ),
+            f,
+        )
+
+
+class ClosingStreamWrapper(IOBase):
+    """A wrapper around a fileobj. It runs cleanup_fn when this wrapper is closed.
+    This is used as a workaround to get around a file closing issue (see open_file).
     """
-    return FileSystems.open(file_path, mode, compression_type)
+
+    def __init__(self, buffer, cleanup_fn, cleanup_args):
+        self._buffer = buffer
+        self._cleanup_fn = cleanup_fn
+        self._cleanup_args = cleanup_args
+
+    def close(self, *args, **kwargs):
+        return self._buffer.close(*args, **kwargs)
+
+    @property
+    def closed(self):
+        return self._buffer.closed
+
+    def fileno(self, *args, **kwargs):
+        return self._buffer.fileno(*args, **kwargs)
+
+    def flush(self, *args, **kwargs):
+        return self._buffer.flush(*args, **kwargs)
+
+    def isatty(self, *args, **kwargs):
+        return self._buffer.isatty(*args, **kwargs)
+
+    def readable(self, *args, **kwargs):
+        return self._buffer.readable(*args, **kwargs)
+
+    def readline(self, *args, **kwargs):
+        return self._buffer.readline(*args, **kwargs)
+
+    def readlines(self, *args, **kwargs):
+        return self._buffer.readlines(*args, **kwargs)
+
+    def seek(self, *args, **kwargs):
+        return self._buffer.seek(*args, **kwargs)
+
+    def seekable(self, *args, **kwargs):
+        return self._buffer.seekable(*args, **kwargs)
+
+    def tell(self, *args, **kwargs):
+        return self._buffer.tell(*args, **kwargs)
+
+    def truncate(self, *args, **kwargs):
+        return self._buffer.truncate(*args, **kwargs)
+
+    def writable(self, *args, **kwargs):
+        return self._buffer.writable(*args, **kwargs)
+
+    def writelines(self, *args, **kwargs):
+        return self._buffer.writelines(*args, **kwargs)
+
+    def read(self, *args, **kwargs):
+        return self._buffer.read(*args, **kwargs)
+
+    def readall(self, *args, **kwargs):
+        return self._buffer.readall(*args, **kwargs)
+
+    def readinto(self, *args, **kwargs):
+        return self._buffer.readinto(*args, **kwargs)
+
+    def write(self, *args, **kwargs):
+        return self._buffer.write(*args, **kwargs)
+
+    def __del__(self, *args, **kwargs):
+        self._cleanup_fn(*self._cleanup_args)
+        return self._buffer.__del__(*args, **kwargs)
+
+
+def open_file(file_path, mode='r'):
+    """
+    Opens the file indicated by the given file path. Can be in a directory.
+    """
+    linked_bundle_path = parse_linked_bundle_url(file_path)
+    if (
+        linked_bundle_path.uses_beam
+        and linked_bundle_path.is_archive
+        and linked_bundle_path.archive_subpath
+    ):
+        # If file path is a .tar.gz file on Azure, open the specified path within the archive.
```

Comment:
Not specific to Azure

Reply:
I think I'm going to keep the Azure references in the comments for now. We don't support anything other than Azure, and I can't think of a good term that's general enough for now, so I think "Azure" is the clearest way to denote the concept we're trying to convey here.

Once we support another storage backend (such as GCP), hopefully then we'll be able to find a more general name (maybe something related to "blob"?), and it would make sense to also substitute the instances of "Azure" with the general name.